### DISCUSSION TITLE
feat: Consider including defsec scanner schema in bundle

### DISCUSSION THREAD

**Discussion by simar7** [aquasecurity/trivy#4197] [Ideas]
### Description

Today defsec scanners have a schema which they need in order to register services. This currently lives here https://github.com/aquasecurity/defsec/blob/d9947704afbd97ed19ab9f34903b33c545621e85/pkg/rego/scanner.go#L122-L131

When new services are added into defsec (via adapters), that requires change to the appropriate schema. For example PR: https://github.com/aquasecurity/defsec/pull/1153 added a new AWS service adapter that required the need for updating the corresponding `cloud.json` schema for AWS.

When the schema is updated (brand new services added) and a new bundle based on that is released, it makes the bundle backwards incompatible, _if_ they use the same MAJOR version. This is the case today as older versions of Trivy will not be aware of the newer services which were added as part of the new bundle.

This causes issues like this: https://github.com/aquasecurity/trivy/issues/4124

#### Short term
When new services are added to defsec, we can update the MAJOR policy version as there is now a significant change (new services being added). This will keep things backwards compatible as older versions of Trivy will not break because of a new policy/service getting added into defsec. The version of defsec policy bundle is pinned here: https://github.com/aquasecurity/trivy/blob/main/pkg/policy/policy.go#L21-L22

This however should only be done less frequently as otherwise (as @knqyf263 pointed out), it will require many Trivy users to frequently update their Trivy version to get the latest updates.

#### Long term
Schema should be treated the same as a policy and should be part of the policy bundle. This will allow Trivy to read the schema of every service from the bundle (if available). We will need to account for cases when bundle is not available (offline mode). For such cases we can consider the same approach we have today for policies, that is to use embedded policies if newer ones are not available, either on intentionally or otherwise.

cc @knqyf263 @itaysk

**Comment by itaysk** [comment]
Conceptually there are these two phases:
1. Trivy reads target configuration from API or file, and Adapter adapts it to defsec common schema.
2. Rego rules are evaluated with that common schema as input.

step1 is shipped with Trivy, step2 is shipped with defsec. Different lifecycles.
The input.schema sits in between these two steps, as it describes the output of step1 for step2.
Your suggestion is to couple the schema with step2, and I just want to verify it doesn't have a dependency with step1.

**Comment by simar7** [comment]
After having taken another look, I think the [improvement to gracefully handle rego errors](https://github.com/aquasecurity/trivy/issues/4183) has helped with this story. This is what happens in different scenarios today:


### Terms
**Correct policy**: Syntactically correct, compiles and works as expected.
**Broken policy**: Syntax issue or a new service that isn't supported in Trivy yet.
**A/B**: In this case version B is newer than version A.

### Case 1: Trivy A + Policy Bundle A (all policies are correct)
Baseline case, everything works as expected.

### Case 2: Trivy A + Policy Bundle A (with broken policies inside)
Trivy continues ahead with the policies that are correct. Errors are logged (on the debug stream) but this does not fatally stop the scanning.

### Case 3: Trivy A + Policy Bundle B (all policies are correct)
Trivy will log errors as Case 2, and will only scan for those services that it has adapters for.

### Case 4: Trivy A + Policy Bundle B (with broken policies inside)
Same as Case 3. Any new policies (that are correct) that are part of this bundle will also be run.

### Case 5: Trivy B + Policy Bundle A
Works as expected as all newer versions of Trivy are compatible with older bundles. Even if we remove a policy for a service in a future Trivy B that used to exist in Policy Bundle A, it will not matter to Trivy B as it will simply not scan it.


### Future Improvements 
Having said that, below is the list of improvements we can make here (add more as I find any):

1. Add an integration test that tests defsec bundles. Today we have added support for loading policy bundles from a custom registry, this functionality can be used to load a bundle and run a scan, like a user would. It will help us avoid failures with bundle incompatibility between Trivy versions.

**Reply by simar7** [reply]
Closing this discussion and tracking the improvement here https://github.com/aquasecurity/trivy/issues/4938

**Reply by itaysk** [reply]
> only scan for those services that it has adapters for

can you please elaborate how you determine this?
Also, how does all of this work with custom policies?

**Reply by simar7** [reply]
> only scan for those services that it has adapters for

Trivy will only scan for those services that it knows about. The newer policies will be logged in the debug output as being unsupported but Trivy will continue to run and scan the rest of the services. This is the current behavior after the recent improvements.

As for custom policies, both cases are covered:
1. __A user provides a custom policy for an existing service__: In this case the policy is evaluated as-is using the schema that Trivy already knows about. A user can also optionally provide a custom schema, if they wish to.
2. __A user provides a custom policy for a new service__: In this case the user also must supply a custom schema to correspond with the policy.
